<launch>
  <arg name="frame_id"/>

  <!-- map information -->
  <arg name="map_size_x_"/>
  <arg name="map_size_y_"/>
  <arg name="map_size_z_"/>

  <!-- <arg name="map_origin_x"/>
  <arg name="map_origin_y"/>
  <arg name="map_origin_z"/> -->


  <!-- drone information -->
  <arg name="mav_id"/>
  
  <arg name="ego_radius" />
  <arg name="ego_height" />

  <arg name="max_vel"/>
  <arg name="max_acc"/>
  <arg name="max_jerk"/>

  <arg name="color/r" />
  <arg name="color/g" />
  <arg name="color/b" />

  <arg name="search/max_tau"/>
  <arg name="planning_horizon"/>


  <!-- topics -->
  <arg name="odom_topic"/>
  <arg name="waypoints_topic"/>
  <arg name="camera_pose_topic"/>
  <arg name="depth_topic"/>
  <arg name="cloud_topic"/>
  <arg name="tracker_topic"/>
  <arg name="poly_srv_name"/>

  <!-- camera parameters -->
  <arg name="cx"/>
  <arg name="cy"/>
  <arg name="fx"/>
  <arg name="fy"/>

  <arg name="off_x"/>
  <arg name="off_y"/>
  <arg name="off_z"/>
  <arg name="qx"/>
  <arg name="qy"/>
  <arg name="qz"/>
  <arg name="qw"/>

  <!-- waypoints -->
  <arg name="point_num"/>
  <arg name="point0_x"/>
  <arg name="point0_y"/>
  <arg name="point0_z"/>
  <arg name="point1_x"/>
  <arg name="point1_y"/>
  <arg name="point1_z"/>
  <arg name="point2_x"/>
  <arg name="point2_y"/>
  <arg name="point2_z"/>

  <arg name="flight_type"/>

  <arg name="initial_offset/x"/>
  <arg name="initial_offset/y"/>


  <!-- main node -->
  <node pkg="plan_manage" name="opt_planner_node" type="opt_planner_node" output="screen">

    <!-- ego informations -->
    <param name="plan_frame_id"  value ="$(arg frame_id)"/>
    <param name="ego_radius"     value="$(arg ego_radius)"/>
    <param name="ego_height"     value="$(arg ego_height)"/>
    <param name="mav_id"         value="$(arg mav_id)"/>

    <param name="color/r"     value="$(arg color/r)"/>
    <param name="color/g"     value="$(arg color/g)"/>
    <param name="color/b"     value="$(arg color/b)"/>
    <param name="color/a"     value="0.1" type="double"/>

    <param name="initial_offset/x"  value="$(arg initial_offset/x)"/>
    <param name="initial_offset/y"  value="$(arg initial_offset/y)"/>
    <param name="poly_srv_name"     value="$(arg poly_srv_name)"/>

    <remap from="~odom_world" to="$(arg odom_topic)"/>
    <remap from="~waypoints"  to="$(arg waypoints_topic)"/>
    <remap from="~tracker_cmd"  to="$(arg tracker_topic)"/>

    <!-- used for occ map -->
    <remap from = "~depth_topic" to = "$(arg depth_topic)"/>

    <param name="max_vel"  value="$(arg max_vel)"  type="double"/>
    <param name="max_acc"  value="$(arg max_acc)"  type="double"/>
    <param name="max_jerk" value="$(arg max_jerk)" type="double"/>

    <!-- planning fsm -->
    <param name="fsm/flight_type" value="$(arg flight_type)" type="int"/>
    <param name="fsm/thresh_replan" value="1.5" type="double"/>
    <param name="fsm/thresh_no_replan" value="2.0" type="double"/>
    <param name="fsm/planning_horizon_time" value="3" type="double"/>
    <param name="fsm/emergency_time_" value="1.0" type="double"/>

    <param name="fsm/waypoint_num" value="$(arg point_num)" type="int"/>
    <param name="fsm/waypoint0_x" value="$(arg point0_x)" type="double"/>
    <param name="fsm/waypoint0_y" value="$(arg point0_y)" type="double"/>
    <param name="fsm/waypoint0_z" value="$(arg point0_z)" type="double"/>
    <param name="fsm/waypoint1_x" value="$(arg point1_x)" type="double"/>
    <param name="fsm/waypoint1_y" value="$(arg point1_y)" type="double"/>
    <param name="fsm/waypoint1_z" value="$(arg point1_z)" type="double"/>
    <param name="fsm/waypoint2_x" value="$(arg point2_x)" type="double"/>
    <param name="fsm/waypoint2_y" value="$(arg point2_y)" type="double"/>
    <param name="fsm/waypoint2_z" value="$(arg point2_z)" type="double"/>
   
    <!--mapping paras global map-->
    <param name="map_size_x" value="$(arg map_size_x_)" /> 
    <param name="map_size_y" value="$(arg map_size_y_)" /> 
    <param name="map_size_z" value="$(arg map_size_z_)" /> 

    <!-- <param name="map_origin_x" value="$(arg map_origin_x)"/>
    <param name="map_origin_y" value="$(arg map_origin_y)"/>
    <param name="map_origin_z" value="$(arg map_origin_z)"/> -->

    <!-- used for occ map -->
    <remap from ="~grid_map_cloud" to="$(arg cloud_topic)"/>
    <remap from = "~grid_map_pose"   to = "$(arg camera_pose_topic)"/> 
    <remap from = "~grid_map_depth" to = "$(arg depth_topic)"/>

    <param name="max_vel"  value="$(arg max_vel)"  type="double"/>
    <param name="max_acc"  value="$(arg max_acc)"  type="double"/>
    <param name="max_jerk" value="$(arg max_jerk)" type="double"/>

    <!-- planning fsm -->
    <param name="fsm/flight_type" value="$(arg flight_type)" type="int"/>
    <param name="fsm/thresh_replan" value="1.5" type="double"/>
    <param name="fsm/thresh_no_replan" value="2.0" type="double"/>
    <param name="fsm/planning_horizon_time" value="3" type="double"/>
    <param name="fsm/emergency_time_" value="1.0" type="double"/>

    <param name="fsm/waypoint_num" value="$(arg point_num)" type="int"/>
    <param name="fsm/waypoint0_x" value="$(arg point0_x)" type="double"/>
    <param name="fsm/waypoint0_y" value="$(arg point0_y)" type="double"/>
    <param name="fsm/waypoint0_z" value="$(arg point0_z)" type="double"/>
    <param name="fsm/waypoint1_x" value="$(arg point1_x)" type="double"/>
    <param name="fsm/waypoint1_y" value="$(arg point1_y)" type="double"/>
    <param name="fsm/waypoint1_z" value="$(arg point1_z)" type="double"/>
    <param name="fsm/waypoint2_x" value="$(arg point2_x)" type="double"/>
    <param name="fsm/waypoint2_y" value="$(arg point2_y)" type="double"/>
    <param name="fsm/waypoint2_z" value="$(arg point2_z)" type="double"/>
   
    <!--mapping paras global map-->
    <param name="map_size_x" value="$(arg map_size_x_)" /> 
    <param name="map_size_y" value="$(arg map_size_y_)" /> 
    <param name="map_size_z" value="$(arg map_size_z_)" /> 

    <!-- <param name="map_origin_x" value="$(arg map_origin_x)"/>
    <param name="map_origin_y" value="$(arg map_origin_y)"/>
    <param name="map_origin_z" value="$(arg map_origin_z)"/> -->

    <param name="grid_map/ground_height" value="-0.1"/>

    <param name="grid_map/resolution"      value="0.1" /> 
    <param name="grid_map/local_update_range_x"  value="6.0" /> 
    <param name="grid_map/local_update_range_y"  value="6.0" /> 
    <param name="grid_map/local_update_range_z"  value="4.0" /> 
    <param name="grid_map/obstacles_inflation"     value="0.1" /> 
    <param name="grid_map/local_map_margin" value="5"/>


    <!-- camera parameter -->
    <param name="grid_map/cx" value="$(arg cx)"/>
    <param name="grid_map/cy" value="$(arg cy)"/>
    <param name="grid_map/fx" value="$(arg fx)"/>
    <param name="grid_map/fy" value="$(arg fy)"/>

    <param name="grid_map/off_x" value="$(arg off_x)"/>
    <param name="grid_map/off_y" value="$(arg off_y)"/>
    <param name="grid_map/off_z" value="$(arg off_z)"/>

    <param name="grid_map/qx" value="$(arg qx)"/>
    <param name="grid_map/qy" value="$(arg qy)"/>
    <param name="grid_map/qz" value="$(arg qz)"/>
    <param name="grid_map/qw" value="$(arg qw)"/>

    <!-- depth filter -->
    <param name="grid_map/use_depth_filter" value="true"/>
    <param name="grid_map/depth_filter_tolerance" value="0.15"/>
    <param name="grid_map/depth_filter_maxdist"   value="5.0"/>
    <param name="grid_map/depth_filter_mindist"   value="0.2"/>
    <param name="grid_map/depth_filter_margin"    value="2"/>
    <param name="grid_map/k_depth_scaling_factor" value="1000.0"/>
    <param name="grid_map/skip_pixel" value="1"/>
    <!-- local fusion -->
    <param name="grid_map/p_hit"  value="0.65"/>
    <param name="grid_map/p_miss" value="0.35"/>
    <param name="grid_map/p_min"  value="0.12"/>
    <param name="grid_map/p_max"  value="0.95"/>
    <param name="grid_map/p_occ"  value="0.80"/>
    <param name="grid_map/min_ray_length" value="0.3"/>
    <param name="grid_map/max_ray_length" value="6.0"/>

    <param name="grid_map/virtual_ceil_height"   value="-5.0"/>
    <param name="grid_map/visualization_truncate_height"   value="3.0"/>
    <param name="grid_map/show_occ_time"  value="false"/>
    <param name="grid_map/pose_type"     value="2"/>  
    <param name="grid_map/frame_id"      value="$(arg frame_id)"/>
    <!--mapping paras end-->

    <!-- kinodynamic path searching -->
    <param name="search/max_tau" value="$(arg search/max_tau)" type="double"/>
    <param name="search/init_max_tau" value="0.5" type="double"/>  
    <param name="search/horizon" value="$(arg planning_horizon)" type="double"/>
    <param name="search/lambda_heu" value="5.0" type="double"/>
    <param name="search/w_time" value="3.0" type="double"/>
    <param name="search/resolution_astar" value="0.1" type="double"/>
    <param name="search/time_resolution" value="0.2" type="double"/>
    <param name="search/margin" value="0.2" type="double"/>
    <param name="search/allocate_num" value="100000" type="int"/>
    <param name="search/vis_check_num" value="15" type="int"/>

    <!-- optimization -->
    <param name="optimization/w_time"     value="32.0" type="double"/>
    <param name="optimization/w_vel"      value="128.0" type="double"/>
    <param name="optimization/w_acc"      value="128.0" type="double"/>
    <param name="optimization/w_sta_obs"  value="4.0" type="double"/>
    <param name="optimization/w_dyn_obs"  value="2048.0" type="double"/>
    <param name="optimization/bb_back"    value="0.8" type="double"/>


  </node>

</launch>
